pkg/parse/listener_impl.go

```diff
@@ -82,10 +86,247 @@ func NewTreeShapeListener() *TreeShapeListener {
 	}
 }
 
+func (s *TreeShapeListener) lint() {
+	s.lintMode = true
+	s.linter = newLinterRecords()
+}
+
 func (s *TreeShapeListener) currentApp() *sysl.Application {
 	return s.app
 }
 
+// linterRecords records any data that is required to do linting
+type linterRecords struct {
+	apps  map[string]*graph
+	calls *graph
+}
+
+type graph map[string]*graphData
+
+type graphData struct {
+	locations map[string]bool
+	rec       *graph
+}
+
+func newAppEndpointGraph() *graph {
+	rec := new(graph)
+	*rec = make(map[string]*graphData)
+	return rec
+}
+
+func newLocation(location string) map[string]bool {
+	loc := make(map[string]bool)
+	loc[location] = true
+	return loc
+}
+
+func (g *graph) recordApp(appName, location string) error {
+	if app, exists := (*g)[appName]; !exists {
+		(*g)[appName] = &graphData{newLocation(location), newAppEndpointGraph()}
+		return nil
+	} else if !app.locations[location] {
+		//TODO: find better way, to handle collector
+		app.locations[location] = true
+		return nil
+	}
+	return fmt.Errorf("recordApp: app already exists: %s %s", location, appName)
```

Comment:
I think it was just meant to handle in case it already exists in the graph, it really shouldn't happen and if it does, that means something is wrong in the implementation.

Reply:
I'll add that in the design docs